File: src/utils/collision.go
package utils

func CollisionWithBbAndPoint(cubic *Cubic, point *Point) bool {
	return point.X > cubic.B && point.X < cubic.D && point.Y > cubic.A && point.Y < cubic.C
}

func CollisionWithBbAndBb(bb1 *Cubic, bb2 *Cubic) bool {
	return CollisionWithBbAndPoint(bb1, &Point{X: bb2.D, Y: bb2.A}) ||
		CollisionWithBbAndPoint(bb1, &Point{X: bb2.B, Y: bb2.A}) ||
		CollisionWithBbAndPoint(bb1, &Point{X: bb2.B, Y: bb2.C}) ||
		CollisionWithBbAndPoint(bb1, &Point{X: bb2.D, Y: bb2.C})
}

File: frontend/src/game/utils/collision.ts
import type { Cubic } from '@root/proto/utils/cubic';
import { Point } from '@root/proto/utils/point';

export class Collision {
    static withBoxAndPoint(cubic: Cubic, point: Point): boolean {
        return (
            point.x > cubic.d &&
            point.x < cubic.b &&
            point.y > cubic.a &&
            point.y < cubic.c
        );
    }

    static withBoxAndBox(cubic1: Cubic, cubic2: Cubic): boolean {
        return (
            Collision.withBoxAndPoint(cubic1, new Point(cubic2.d, cubic2.a)) ||
            Collision.withBoxAndPoint(cubic1, new Point(cubic2.b, cubic2.a)) ||
            Collision.withBoxAndPoint(cubic1, new Point(cubic2.b, cubic2.c)) ||
            Collision.withBoxAndPoint(cubic1, new Point(cubic2.d, cubic2.c))
        );
    }
}
